src/Type/ArrayType.php

```diff
@@ -286,9 +286,11 @@ public function setOffsetValueType(?Type $offsetType, Type $valueType, bool $uni
 
 		$array = new self(
 			TypeCombinator::union($this->keyType, $offsetType),
-			$unionValues ? TypeCombinator::union($this->itemType, $valueType) : $valueType,
+			$unionValues || $offsetType instanceof ConstantStringType || $offsetType instanceof ConstantIntegerType
```

Comment:
I don't love this fix, what sense does it make? Why do you choose to union the values even for `!$unionValues`?

Reply:
When you have the following kind of array
```
non-empty-array<string, mixed>&hasOffsetValue('uses', array)
```
And when you're doing
```
$array['uses'][] = '';
```
Phpstan is trying to do
```
$valueToWrite = $offsetValueType->setOffsetValueType($offsetType, $valueToWrite, $i === 0);
```
where `$i` is 1, so unionValues is false.

`setOffsetValueType` is called on an intersection Type so it's applied on every types
`hasOffsetValue('uses', array)` become `hasOffsetValue('uses', non-empty-array)`
and `non-empty-array<string, mixed>` become `non-empty-array<string, non-empty-array>` (which is wrong)

I wanted that `non-empty-array<string, mixed>` stay like it was. So I thought that when you're calling setOffsetValue with a specific constant key, the generic array type should still keep the union strategy. (in this case mixed will stay mixed).

But maybe something should be done to the `IntersectionType::offsetValue` method.
Something like, if `there is `hasOffsetValueType`, the `ArrayType` must not call `setOffsetValueType`...

Reply:
In fact the issue is that
```
$array['uses'][] = '';
```
and 
```
$array['uses'] = [...$array['uses'], ''];
```
doesn't behave the same way.